torch/csrc/autograd/profiler.cpp

```diff
@@ -93,27 +93,126 @@ void popRange() {
   }
 }
 
-RecordFunction::RecordFunction(Function* fn) {
-  // typeid(*fn).name() would avoid an additional string allocation.
-  // However, typeid(*fn).name() would cause nvtx annotations for all user-defined
-  // (Python-side) custom autograd function backward() methods to have the same name,
-  // because they route through the same C++ side class.
-  // fn->name() ensures that nvtx annotations for custom function backward() methods
-  // receive a relevant, demangled name.
+// typeid(*fn).name() would avoid an additional string allocation.
+// However, typeid(*fn).name() would cause nvtx annotations for all user-defined
+// (Python-side) custom autograd function backward() methods to have the same name,
+// because they route through the same C++ side class.
+// fn->name() ensures that nvtx annotations for custom function backward() methods
+// receive a relevant, demangled name.
+FunctionCallContext::FunctionCallContext(Function* fn)
+    : fn_(fn), owned_name_(new std::string(fn->name())), name_ptr_(owned_name_->c_str()),
+      sequence_nr_(fn->sequence_nr()) {}
+
+FunctionCallContext::FunctionCallContext(
+    std::string name, int64_t sequence_nr)
+    : owned_name_(new std::string(std::move(name))), name_ptr_(owned_name_->c_str()),
+      sequence_nr_(sequence_nr) {}
+
+FunctionCallContext::FunctionCallContext(
+    const char* name_ptr, int64_t sequence_nr)
+    : name_ptr_(name_ptr), sequence_nr_(sequence_nr) {}
+
+
+namespace {
+std::mutex callbacks_mutex;
+std::atomic<RecordCallbacks*> global_callbacks(nullptr);
```

Comment:
I don't get why does this have to be thread safe if the profiler in itself isn't? We can save on atomic load/stores and mutex locks if we get rid of this requirement.

Reply:
because we'd want to be able to add/remove callbacks while the program is running